saleor/checkout/utils.py

```diff
@@ -1070,3 +1070,20 @@ def get_address_for_checkout_taxes(
 ) -> Optional["Address"]:
     shipping_address = checkout_info.delivery_method_info.shipping_address
     return shipping_address or checkout_info.billing_address
+
+
+def save_checkout_if_not_deleted(
+    checkout, fields_to_update: Optional[list] = None
+) -> None:
+    with transaction.atomic():
+        # check if checkout still exists and lock item for updates
+        if not Checkout.objects.select_for_update().filter(pk=checkout.pk).first():
+            raise ValidationError(
+                {
+                    "checkout": ValidationError(
+                        "Checkout does no longer exists.",
+                        code=CheckoutErrorCode.DELETED.value,
+                    )
+                }
+            )
+        checkout.save(**{"update_fields": fields_to_update} if fields_to_update else {})
```

Comment:
What's the benefit of locking a row immediately before updating it? You could do a savepoint, attempt an update, and if unsuccessful, rollback to the savepoint. Doing it this way is costly and unnecessarily fetches the entire row from the database.

Reply:
My first implementation was similar to what you say - save an object and if `DatabaseError` occurs, raise an error - no locking there. There was a check to filter only `DatabaseError` of certain types and message (I couldn't find a better way of filtering that error).